src/Widgets/TaskRow.vala

```diff
@@ -67,25 +66,97 @@ public class Tasks.TaskRow : Gtk.ListBoxRow {
         grid.attach (summary_label, 1, 0);
         grid.attach (description_grid, 1, 1);
 
-        if ( !ical_task.get_due ().is_null_time () ) {
+        due_label = new Gtk.Label (null);
+        unowned Gtk.StyleContext due_label_context = due_label.get_style_context ();
+        due_label_context.add_provider (taskrow_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+        due_label_context.add_class ("due-date");
+        description_grid.add (due_label);
+
+        description_label = new Gtk.Label (null);
+        description_label.xalign = 0;
+        description_label.lines = 1;
+        description_label.ellipsize = Pango.EllipsizeMode.END;
+        description_label.get_style_context ().add_class (Gtk.STYLE_CLASS_DIM_LABEL);
+        description_grid.add (description_label);
+
+        var eventbox = new Gtk.EventBox ();
+        eventbox.expand = true;
+        eventbox.above_child = false;
+        eventbox.add (grid);
+
+        add (eventbox);
+
+        eventbox.event.connect ((event) => {
+            if (event.type == Gdk.EventType.@2BUTTON_PRESS) {
+                var task_popover = new Tasks.TaskSettingsPopover (task);
+                task_popover.constrain_to = Gtk.PopoverConstraint.NONE;
+                task_popover.position = Gtk.PositionType.LEFT;
+                task_popover.set_relative_to (check);
+
+                task_popover.closed.connect (() => {
+                    task_changed (task);
+                });
+
+                task_popover.popup ();
+
+            } else if (!is_selected () && event.type == Gdk.EventType.BUTTON_PRESS) {
+                activate ();
+            }
+        });
+
+        check.toggled.connect (() => {
+            if (task == null) {
+                return;
+            }
+            task.get_icalcomponent ().set_status (check.active ? ICal.PropertyStatus.COMPLETED : ICal.PropertyStatus.NONE);
+            task_changed (task);
+        });
+
+        key_release_event.connect ((event) => {
+            // SPACE == 32
+            if (event.keyval == 32) {
+                check.active = !check.active;
+            }
+        });
+    }
+
+    public void update_request (ECal.Component task) {
+        this.task = task;
```

Comment:
Instead of doing things this way around, I think I would probably have task be a `public { get; construct set; }` and then add a `notify["task"].connect ()`. I think it feels more consistent to set object properties using properties instead of methods :)

Reply:
good catch! I really wanted it to be this way, but wasn't aware that we can combine `construct set;` - I assumed it's an either/or. Will update the code to reflect this!